pkg/sql/virtual_schema.go

```diff
@@ -698,16 +743,72 @@ func (e *virtualDefEntry) makeConstrainedRowsGenerator(
 			newConstraint.Spans.Append(idxConstraint.Spans.Get(currentSpan))
 		}
 
+		// We'd want to walk the expression, figure out the columns we need, then
+		{
+			var err error
+			filter, err = e.maybeGetIndexPredicateFilter(ctx, p, index)
+			if err != nil {
+				return err
+			}
+		}
+
 		// NB: If we allow virtualSchemaTables with generator to perform a constrained scan,
 		// we then need to ensure that we don't call populate without checking, as it may be nil.
 		if def.populate == nil {
 			return errors.AssertionFailedf(
 				"programming error: can't fall back to unconstrained scan on generated vtables")
 		}
+
 		return def.populate(ctx, p, dbDesc, addRowIfPassesFilter(&newConstraint))
 	}
 }
 
+// maybeGetIndexPredicateFilter returns a function which can be used to filter
+// rows of a virtual table which do not match the corresponding index predicate.
+// It will return nil if the index is nil or is not partial. We need this
+// because there are cases the optimizer will choose to scan a virtual index
+// but the index cannot be used to serve the query. Instead, we need to scan
+// the primary index and then constrain it as though the partial index were
+// scanned.
+func (e *virtualDefEntry) maybeGetIndexPredicateFilter(
+	ctx context.Context, p *planner, index catalog.Index,
+) (func(datums tree.Datums) (matched bool, _ error), error) {
+	if index == nil {
+		return nil, nil
+	}
+	pred := index.GetPredicate()
+	if pred == "" {
+		return nil, nil
+	}
+	publicColumns := e.desc.PublicColumns()
+	exprs, _, err := schemaexpr.MakePartialIndexExprs(
```

Comment:
nit: make a function that can make a single partial index expression, or a leave a TODO for me here

Reply:
Done.